src/trs/rewrite.rs

```diff
@@ -309,6 +309,152 @@ impl TRS {
+    /// # use rand::{thread_rng};
+    /// # use term_rewriting::{Context, RuleContext, Signature, parse_rule};
+    /// # fn main() {
+    /// let mut sig = Signature::default();
+    ///
+    /// let rule = parse_rule(&mut sig, "A(x_) = B(x_)").expect("parse of A(x_) = B(x_)");
+    ///
+    /// let new_rule = TRS::swap_lhs_and_one_rhs_helper(&rule);
+    ///
+    /// if new_rule == None {
+    ///     assert!(false);
+    /// } else {
+    ///     assert_eq!(new_rule.unwrap().display(&sig), "B(x_) = A(x_)");
+    /// }
+    /// # }
+    /// ```
+    pub fn swap_lhs_and_one_rhs_helper(rule: &Rule) -> Option<Rule> {
+        let r = rule.clone();
+        let rhs = match r.rhs() {
+            Some(rh) => rh,
+            None => {
+                return None;
+            }
+        };
+        if rhs.variables().len() == r.lhs.variables().len() {
+            let new_rhs = vec![r.lhs];
+            return Rule::new(rhs, new_rhs);
+        }
+        return None;
+    }
+    /// returns a vector of a rules with each rhs being the lhs of the original
+    /// rule and each lhs is each rhs of the original.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # #[macro_use] extern crate polytype;
+    /// # extern crate programinduction;
+    /// # extern crate rand;
+    /// # extern crate term_rewriting;
+    /// # extern crate itertools;
+    /// # use programinduction::trs::{TRS, Lexicon};
+    /// # use rand::{thread_rng};
+    /// # use term_rewriting::{Context, RuleContext, Signature, parse_rule};
+    /// # use itertools::Itertools;
+    /// # fn main() {
+    /// let mut sig = Signature::default();
+    ///
+    /// let rule = parse_rule(&mut sig, "A(x_) = B(x_) | C(x_)").expect("parse of A(x_) = B(x_) | C(x_)");
+    ///
+    /// let new_rules = TRS::swap_lhs_and_all_rhs_helper(&rule);
+    /// if new_rules == None {
+    ///     assert!(false);
+    /// } else {
+    ///     let rules = new_rules.unwrap().iter().map(|r| format!("{};", r.display(&sig))).join("\n");
+    ///     assert_eq!(rules, "B(x_) = A(x_);\nC(x_) = A(x_);");
+    /// }
+    /// # }
+    /// ```
+    pub fn swap_lhs_and_all_rhs_helper(rule: &Rule) -> Option<Vec<Rule>> {
+        let mut rules: Vec<Rule> = vec![];
+        let num_vars = rule.variables().len();
+        for idx in 0..rule.len() {
+            if rule.rhs[idx].variables().len() == num_vars {
+                let lhs = rule.rhs[idx].clone();
+                let rhs = vec![rule.lhs.clone()];
+                let temp_rule = Rule::new(lhs, rhs);
+                if temp_rule != None {
+                    rules.push(temp_rule.unwrap());
+                }
+            }
+        }
+        if rules.len() == 0 {
+            return None;
+        }
+        return Some(rules);
+    }
+    /// Selects a rule from the TRS at random, swaps the LHS and RHS if possible and inserts the resulting rules
+    /// back into the TRS imediately after the background.
```

Comment:
Can you please add a test here and for `swap_lhs_and_rhs_vec` to give people some idea of how they work?

Reply:
Great - thanks for the new test. Can you also, as with the others, add cases that result in `Err` rather than `Ok`, both here and for `swap_lhs_and_rhs_vec`?

Reply:
So right now I have it returning the original TRS if there is an issue with the swap, should it fail and Error out or should we keep it as it is?

Reply:
I'd prefer to see it return an `Err`. Otherwise, we don't need to return a `Result`, but `Result` provides useful information: did what I was trying to do (i.e. generate a new `TRS`) succeed or fail? That gets lost if you silently return the original and pretend that it's somehow new.

Reply:
Should I create a new Err type for this specific error or should I just use a pre - existing one such as the options exhausted variant?

Reply:
options exhausted is a fine choice.